src/server_manager/web_app/ui_components/outline-per-key-data-limit-dialog.ts

```diff
@@ -0,0 +1,380 @@
+        </div>
+        <div id="menuSection">
+          <paper-checkbox ?checked=${this._showMenu} @tap=${this._setCustomLimitTapped}>
+            ${this.localize('per-key-data-limit-dialog-set-custom')}
+          </paper-checkbox>
+          ${this._showMenu ? this.renderMenu() : ''}
+        </div>
+        <div id="buttonsSection">
+          <paper-button id="save" ?disabled=${!this._enableSave} @tap=${this._sendSaveEvent}>${
+        this.localize('save')}</paper-button>
+          <paper-button @tap=${this.close}>${this.localize('cancel')}</paper-button>
+        </div>
+      </paper-dialog>
+    `;
+  }
+
+  private renderMenu() {
+    return html`
+      <div id="menu">
+        <paper-input
+          id="dataLimitInput"
+          label=${this.localize('data-limit')}
+          always-float-label
+          allowed-pattern="[0-9\\.]"
+          pattern="[0-9]+(\\.[0-9]*)?"
+          auto-validate
+          value=${this._initialValue()}
+          size="7"
+          @keyup=${this._setSaveButtonDisabledState}
+        >
+        </paper-input>
+        <paper-dropdown-menu id="unitsDropdown" noink>
+          <paper-listbox
+            id="unitsListbox"
+            slot="dropdown-content"
+            attr-for-selected="name"
+            selected="${this._initialUnit()}"
+          >
+            <paper-item name="GB">${this._getInternationalizedUnit(1000000000)}</paper-item>
+            <paper-item name="MB">${this._getInternationalizedUnit(1000000)}</paper-item>
+          </paper-listbox>
+        </paper-dropdown-menu>
+      </div>
+    `;
+  }
+
+  private _queryAs<T extends HTMLElement>(selector: string): T {
+    return this.shadowRoot.querySelector(selector) as T;
+  }
+
+  private get _input(): PaperInputElement {
+    return this._queryAs<PaperInputElement>('#dataLimitInput');
+  }
+
+  private _dataLimitValue() {
+    return Number(this._input?.value) ?? 0;
+  }
+
+  private _dataLimitType() {
+    return this._queryAs<PaperListboxElement>('#unitsListbox').selected as 'GB' | 'MB';
+  }
+
+  private _getInternationalizedUnit(bytes: number) {
+    return formatBytesParts(bytes, this._language).unit;
+  }
+
+  private _initialUnit(): 'GB'|'MB' {
+    return bytesToDisplayDataAmount(this._activeDataLimit())?.unit || 'GB';
+  }
+
+  private _initialValue() {
+    return bytesToDisplayDataAmount(this._activeDataLimit())?.value.toString() || '';
+  }
+
+  private _activeDataLimit(): number|undefined {
+    // Returns the limit which currently is enforced on this key, or undefined if there is none.
+    return this._keyDataLimitBytes ?? this._serverDefaultLimitBytes;
+  }
+
+  private async _setCustomLimitTapped() {
+    this._showMenu = !this._showMenu;
+    if (this._showMenu) {
+      await this.updateComplete;
+      this._input?.focus();
+    }
+  }
+
+  private _setSaveButtonDisabledState() {
+    this._enableSave = !this._input?.invalid ?? false;
```

Comment:
`_enableSave` needs to be a property so that we can listen to it and trigger a render if the input is invalid.  Something like 1.2.3 is allowed to by typed by the `allowed-pattern` regex since it checks character-by-character, not the full input string.